Framework/CurveFitting/src/Functions/Meier.cpp

```diff
@@ -0,0 +1,184 @@
+
+  std::valarray<double> positiveLambda;
+  std::valarray<double> negativeLambda;
+  std::valarray<double> cos2AlphaSquared;
+  std::valarray<double> sin2AlphaSquared;
+  std::valarray<double> cosAlphaSquared;
+  std::valarray<double> sinAlphaSquared;
+
+  precomputeIntermediateSteps(sinAlphaSquared, cosAlphaSquared, sin2AlphaSquared, cos2AlphaSquared, negativeLambda,
+                              positiveLambda, J2);
+
+  std::valarray<double> Px;
+  calculatePx(Px, xValArray, sinAlphaSquared, cosAlphaSquared, negativeLambda, positiveLambda, J2);
+
+  std::valarray<double> Pz;
+  calculatePz(Pz, xValArray, sin2AlphaSquared, cos2AlphaSquared, negativeLambda, positiveLambda, J2);
+
+  const std::valarray<double> outValArray = A0 * gau * Lor * (1. / 3.) * (2 * Px + Pz);
+  std::copy(begin(outValArray), end(outValArray), out);
+}
+
+/**
+ * Precomputes intermediate terms used to calculate the polrization in the x and z directions. All value arrays will be
+ * resized to J2 + 2 and set to their respective quantities
+ * @param sinAlphaSquared :: sin of alpha squared
+ * @param cosAlphaSquared :: cos of alpha squared
+ * @param sin2AlphaSquared :: sin of 2*alpha squared
+ * @param cos2AlphaSquared :: cos of 2*alpha squared
+ * @param positiveLambda :: negative lambda
+ * @param negativeLambda :: positive lambda
+ * @param J2 :: 2 * total angular momentum quantum number
+ */
+void Meier::precomputeIntermediateSteps(std::valarray<double> &sinAlphaSquared, std::valarray<double> &cosAlphaSquared,
+                                        std::valarray<double> &sin2AlphaSquared,
+                                        std::valarray<double> &cos2AlphaSquared, std::valarray<double> &negativeLambda,
+                                        std::valarray<double> &positiveLambda, const double &J2) const {
+  const double FreqD = getParameter("FreqD");
+  const double FreqQ = getParameter("FreqQ");
+  const double J = J2 / 2;
+
+  const double OmegaD = 2 * M_PI * FreqD;
+  const double OmegaQ = 2 * M_PI * FreqQ;
+
+  const size_t size = int(J2 + 2);
+  sinAlphaSquared.resize(size);
+  cosAlphaSquared.resize(size);
+  sin2AlphaSquared.resize(size);
+  cos2AlphaSquared.resize(size);
+  negativeLambda.resize(size);
+  positiveLambda.resize(size);
+  std::valarray<double> Wm(size);
+
+  for (size_t i = 0; i < size; i++) {
+    const double m = static_cast<double>(i) - J;
+    const double q1 = (OmegaQ + OmegaD) * (2 * m - 1);
+    const double q2 = OmegaD * std::sqrt(J * (J + 1) - m * (m - 1));
+    const double qq = pow(q1, 2) + pow(q2, 2);
+    const double q3 = OmegaQ * (2 * pow(m, 2) - 2 * m + 1) + OmegaD;
+
+    Wm[i] = std::sqrt(qq);
+
+    if (static_cast<double>(i) < (J2 + 1)) {
+      positiveLambda[i] = 0.5 * (q3 + Wm[i]);
+    } else {
+      positiveLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;
+    }
+
+    if (i > 0) {
+      negativeLambda[i] = 0.5 * (q3 - Wm[i]);
+    } else {
+      negativeLambda[i] = OmegaQ * pow(J, 2) - OmegaD * J;
```

Comment:
we only use this for the first iteration of the loop since i starts at 0. Could we not extract the if statements into functions and then calculate the i=0 case before the for loop? I think this would remove the need for this if statement.

Reply:
Yes, but I don't think this will improve the readability of the code

Reply:
this is more to do with performance than readability. If statements are slow and these are in a for loop. Then it will be called from a fit which may evaluate the function 1000's of times. So the split second longer one if takes will add up when repeated so many times.

Reply:
ok